(*Iter).MapScan fails on tuples in Cassandra 2.2.4

The program panics on the first call to MapScan() due to being unable to reflect.New(nil). Debugging suggests it consideres the TupleType as custom, as the func "getApacheCassandraType(class string) (Type)" receives "org.apache.cassandra.db.marshal.TupleType(org.apache.cassandra.db.marshal.Int32Type,org.apache.cassandra.db.marshal.Int32Type)" in its "class" argument.

The method continues to strips it down to: 
"TupleType(org.apache.cassandra.db.marshal.Int32Type,org.apache.cassandra.db.marshal.Int32Type)" before comparing it to every type, including "TupleType" in a switch statement. But the tuple-contents prevents it from recognizing it. Thus we end up with a panic when the type is custom (=0x00 / nil).

The following is a code example reproducing the error, I appreciate any help on this matter:
``` go
package main

import (
	"fmt"
	"log"

	"github.com/gocql/gocql"
)

func main() {
	cluster := gocql.NewCluster("127.0.0.1")
	cluster.Consistency = gocql.One

	session, err := cluster.CreateSession()
	if err != nil {
		log.Fatalln(err)
	}

    // Create testing keyspace
    if err := session.Query(`CREATE KEYSPACE IF NOT EXISTS gocqldebugging_test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};`).Exec(); err != nil {
        log.Fatalln(err)
    }

    // Create table with tuple in it
    if err := session.Query(`CREATE TABLE IF NOT EXISTS gocqldebugging_test.test (id uuid PRIMARY KEY, myNonCustom frozen<tuple<int, int>>);`).Exec(); err != nil {
        log.Fatalln(err)
    }

    // Insert some data into our table
    if err := session.Query(`INSERT INTO gocqldebugging_test.test (id, myNonCustom) VALUES (uuid(), (1, 2));`).Exec(); err != nil {
        log.Fatalln(err)
    }

    fmt.Println("Results:")
	var m map[string]interface{}
	iter := session.Query(`SELECT * FROM gocqldebugging_test.test`).Iter()
	for iter.MapScan(m) {
		fmt.Printf("%T: %#v\n", m, m)
	}

	fmt.Println("Done.")
}

```

And the panic message (brief; test to see full):

```bash
panic: reflect: New(nil)

goroutine 1 [running]:
reflect.New(0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.2/libexec/src/reflect/value.go:2113 +0x85
```

--- a/cassandra_test.go
+++ b/cassandra_test.go
@@ -336,8 +336,8 @@ func TestCAS(t *testing.T) {
 	}
 
 	if _, err := session.Query(`DELETE FROM cas_table WHERE title = ? and revid = ? IF last_modified = ?`,
-		title, revid, tenSecondsLater).ScanCAS(); err.Error() != "count mismatch" {
-		t.Fatalf("delete: was expecting count mismatch error but got %s", err)
+		title, revid, tenSecondsLater).ScanCAS(); !strings.HasPrefix(err.Error(), "gocql: not enough columns to scan into") {
+		t.Fatal("delete: was expecting count mismatch error but got: %q", err.Error())
 	}
 
 	if applied, err := session.Query(`DELETE FROM cas_table WHERE title = ? and revid = ? IF last_modified = ?`,

--- a/helpers.go
+++ b/helpers.go
@@ -5,6 +5,7 @@
 package gocql
 
 import (
+	"fmt"
 	"math/big"
 	"reflect"
 	"strings"
@@ -128,16 +129,34 @@ func (r *RowData) rowMap(m map[string]interface{}) {
 	}
 }
 
+// TupeColumnName will return the column name of a tuple value in a column named
+// c at index n. It should be used if a specific element within a tuple is needed
+// to be extracted from a map returned from SliceMap or MapScan.
+func TupleColumnName(c string, n int) string {
+	return fmt.Sprintf("%s[%d]", c, n)
+}
+
 func (iter *Iter) RowData() (RowData, error) {
 	if iter.err != nil {
 		return RowData{}, iter.err
 	}
+
 	columns := make([]string, 0)
 	values := make([]interface{}, 0)
+
 	for _, column := range iter.Columns() {
-		val := column.TypeInfo.New()
-		columns = append(columns, column.Name)
-		values = append(values, val)
+
+		switch c := column.TypeInfo.(type) {
+		case TupleTypeInfo:
+			for i, elem := range c.Elems {
+				columns = append(columns, TupleColumnName(column.Name, i))
+				values = append(values, elem.New())
+			}
+		default:
+			val := column.TypeInfo.New()
+			columns = append(columns, column.Name)
+			values = append(values, val)
+		}
 	}
 	rowData := RowData{
 		Columns: columns,

--- a/marshal.go
+++ b/marshal.go
@@ -1616,6 +1616,10 @@ type TupleTypeInfo struct {
 	Elems []TypeInfo
 }
 
+func (t TupleTypeInfo) New() interface{} {
+	return reflect.New(goType(t)).Interface()
+}
+
 type UDTField struct {
 	Name string
 	Type TypeInfo
@@ -1626,6 +1630,10 @@ type UDTTypeInfo struct {
 	KeySpace string
 	Name     string
 	Elements []UDTField
+}
+
+func (u UDTTypeInfo) New() interface{} {
+	return reflect.New(goType(u)).Interface()
 }
 
 func (u UDTTypeInfo) String() string {

--- a/session.go
+++ b/session.go
@@ -834,7 +834,7 @@ func (iter *Iter) Scan(dest ...interface{}) bool {
 	// currently only support scanning into an expand tuple, such that its the same
 	// as scanning in more values from a single column
 	if len(dest) != iter.meta.actualColCount {
-		iter.err = errors.New("count mismatch")
+		iter.err = fmt.Errorf("gocql: not enough columns to scan into: have %d want %d", len(dest), iter.meta.actualColCount)
 		return false
 	}
 

--- a/tuple_test.go
+++ b/tuple_test.go
@@ -49,3 +49,31 @@ func TestTupleSimple(t *testing.T) {
 		t.Errorf("expected to get coord.y=-100 got: %v", coord.y)
 	}
 }
+
+func TestTupleMapScan(t *testing.T) {
+	if *flagProto < protoVersion3 {
+		t.Skip("tuple types are only available of proto>=3")
+	}
+
+	session := createSession(t)
+	defer session.Close()
+
+	err := createTable(session, `CREATE TABLE gocql_test.tuple_map_scan(
+		id int,
+		val frozen<tuple<int, int>>,
+
+		primary key(id))`)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err := session.Query(`INSERT INTO tuple_map_scan (id, val) VALUES (1, (1, 2));`).Exec(); err != nil {
+		t.Fatal(err)
+	}
+
+	m := make(map[string]interface{})
+	err = session.Query(`SELECT * FROM tuple_map_scan`).MapScan(m)
+	if err != nil {
+		t.Fatal(err)
+	}
+}

